controllers/flp/flp_controller.go

```diff
@@ -0,0 +1,228 @@
+	"github.com/netobserv/network-observability-operator/pkg/loki"
+	"github.com/netobserv/network-observability-operator/pkg/manager"
+	"github.com/netobserv/network-observability-operator/pkg/manager/status"
+	"github.com/netobserv/network-observability-operator/pkg/watchers"
+	configv1 "github.com/openshift/api/config/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	ascv2 "k8s.io/api/autoscaling/v2"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Reconciler reconciles the current flowlogs-pipeline state with the desired configuration
+type Reconciler struct {
+	client.Client
+	mgr       *manager.Manager
+	watcher   *watchers.Watcher
+	status    status.Instance
+	clusterID string
+}
+
+func Start(ctx context.Context, mgr *manager.Manager) error {
+	log := log.FromContext(ctx)
+	log.Info("Starting Flowlogs Pipeline parent controller")
+
+	r := Reconciler{
+		Client: mgr.Client,
+		mgr:    mgr,
+		status: mgr.Status.ForComponent(status.FLPParent),
+	}
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&flowslatest.FlowCollector{}, reconcilers.IgnoreStatusChange).
+		Named("flp").
+		Owns(&appsv1.Deployment{}).
+		Owns(&appsv1.DaemonSet{}).
+		Owns(&ascv2.HorizontalPodAutoscaler{}).
+		Owns(&corev1.Namespace{}).
+		Owns(&corev1.Service{}).
+		Owns(&corev1.ServiceAccount{})
+
+	ctrl, err := builder.Build(&r)
+	if err != nil {
+		return err
+	}
+	r.watcher = watchers.NewWatcher(ctrl)
+
+	return nil
+}
+
+type subReconciler interface {
+	context(ctx context.Context) context.Context
+	cleanupNamespace(ctx context.Context)
+	reconcile(ctx context.Context, desired *flowslatest.FlowCollector) error
+	getStatus() *status.Instance
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	l := log.Log.WithName("flp") // clear context (too noisy)
+	ctx = log.IntoContext(ctx, l)
+
+	r.status.SetUnknown()
+	defer r.status.Commit(ctx, r.Client)
+
+	err := r.reconcile(ctx)
+	if err != nil {
+		l.Error(err, "FLP reconcile failure")
+		// Set status failure unless it was already set
+		if !r.status.HasFailure() {
+			r.status.SetFailure("FLPError", err.Error())
+		}
+		return ctrl.Result{}, err
+	}
+
+	r.status.SetReady()
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) reconcile(ctx context.Context) error {
+	log := log.FromContext(ctx)
+
+	clh, fc, err := helper.NewFlowCollectorClientHelper(ctx, r.Client)
+	if err != nil {
+		return fmt.Errorf("failed to get FlowCollector: %w", err)
+	} else if fc == nil {
+		return nil
+	}
+
+	ns := helper.GetNamespace(&fc.Spec)
+	previousNamespace := r.status.GetDeployedNamespace(fc)
+	loki := helper.NewLokiConfig(&fc.Spec.Loki)
+	cmn := r.newCommonInfo(clh, ns, previousNamespace, &loki)
+
+	r.watcher.Reset(ns)
+
+	// obtain default cluster ID - api is specific to openshift
+	if r.mgr.IsOpenShift() && r.clusterID == "" {
```

Comment:
May be we can move it to the start function ?

Reply:
I also tried that :) and it doesn't work because  in `Start` the cache hasn't started yet, so all client calls fail

Reply:
(Given it's cached, it's a negligible performance penalty)